### ISSUE TITLE
`UNRESOLVED_ENV_VAR` reported when add external rust file with `include!`

### ISSUE THREAD

**Issue by fre3dm4n** [rust-lang/rust-analyzer#17593]
The code is at [repo for reproduce](https://github.com/fre3dm4n/rust-analyzer-issue), you can build it with 
```bash
BINDING_PATH=../../build/bindings.rs cargo build --manifest-path ./proj/Cargo.toml
```

However when editing in VSCode, rust-analyzer report `UNRESOLVED_ENV_VAR` on the `include!` macro:

![image](https://github.com/user-attachments/assets/2cf6c74c-af35-45f0-8b19-69f0f70aba4f)

I think I already set necessary variables in the `settings.json`
```rs
{
    "rust-analyzer.cargo.extraEnv": {
        "BINDING_PATH": "${workspaceFolder}/build/bindings.rs"
    },
    "rust-analyzer.check.extraEnv": {
        "BINDING_PATH": "${workspaceFolder}/build/bindings.rs"
    },
    "rust-analyzer.restartServerOnConfigChange": true,
}

```

How can I eliminate this error and enable IntelliSense for the `bindings` crate?

**rust-analyzer version**:  `rust-analyzer version: 0.3.2029-standalone`

**rustc version**: `rustc 1.79.0 (129f3b996 2024-06-10)`

**editor or extension**: VSCode

**relevant settings**: null

**repository link (if public, optional)**: [repo for reproduce](https://github.com/fre3dm4n/rust-analyzer-issue)

**code snippet to reproduce**: 
You can find the code at the repository above

**Comment by lnicola** [comment]
For these we currently support paths relative to `OUT_DIR`.

**Comment by fre3dm4n** [comment]
Oh, I see. I should read the source code myself. Is this a thoughtful design choice or something that could be improved in the future? Supporting only `OUT_DIR` seems quite insufficient in certain cases.

**Comment by fre3dm4n** [comment]
Hi @lnicola, any help? Thanks a lot 🫠

**Comment by lnicola** [comment]
Maybe you can say more about your use case?

**Comment by fre3dm4n** [comment]
> Maybe you can say more about your use case?

In my case, I'm trying to integrate Rust code into a C repository that uses Meson as its build system. Meson has its own method for generating Rust bindings for C, so the generated files are not placed in Rust's `OUT_DIR`.

**Comment by fre3dm4n** [comment]
Perhaps allowing the user to provide some environment settings for Rust Analyzer's diagnostics?

**Comment by davidbarsky** [comment]
> > Maybe you can say more about your use case?
> 
> In my case, I'm trying to integrate Rust code into a C repository that uses Meson as its build system. Meson has its own method for generating Rust bindings for C, so the generated files are not placed in Rust's `OUT_DIR`.

Makes sense, thanks! Do you mind linking me to where Meson places generated code? I haven't tested this, but the following might work:

```json
"rust-analyzer.server.extraEnv": {
        "OUT_DIR": "${workspaceFolder}/build/bindings.rs"
},
```

...and changing the `include!()` to use `OUT_DIR` instead.

I don't recall if `${workspaceFolder}` will be interpolated by VS Code—I don't believe it will—so if that doesn't work, I'd try an absolute path.

**Comment by fre3dm4n** [comment]
> > > Maybe you can say more about your use case?
> > 
> > 
> > In my case, I'm trying to integrate Rust code into a C repository that uses Meson as its build system. Meson has its own method for generating Rust bindings for C, so the generated files are not placed in Rust's `OUT_DIR`.
> 
> Makes sense, thanks! Do you mind linking me to where Meson places generated code? I haven't tested this, but the following might work:
> 
> ```json
> "rust-analyzer.server.extraEnv": {
>         "OUT_DIR": "${workspaceFolder}/build/bindings.rs"
> },
> ```
> 
> ...and changing the `include!()` to use `OUT_DIR` instead.
> 
> I don't recall if `${workspaceFolder}` will be interpolated by VS Code—I don't believe it will—so if that doesn't work, I'd try an absolute path.

I've added some code to the repository to help reproduce this issue. Please check it out to see if it helps.

[rust-analyzer-issue](https://github.com/fre3dm4n/rust-analyzer-issue)

**Comment by msandy** [comment]
FWIW I'm planning to use this functionality in my project's code. It seems like regardless of what best-practices or idiomatic rust might be, `rust-analyzer` should align to whatever is supported in `rustc`. In this case, it is fully supported, and so should be supported by `rust-analyzer` IMHO. Having behavioral divergence between the language server and the compiler just leads to fragmentation.
